plonky2/src/plonk/circuit_builder.rs

```diff
@@ -392,6 +397,48 @@ impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
         })
     }
 
+    /// Find an available slot, of the form `(gate_index, op)` for gate `G`.
+    pub fn find_slot<G: Gate<F, D> + Clone>(
+        &mut self,
+        gate: G,
+        params: &[F],
+        constants: &[F],
+    ) -> (usize, usize) {
+        let num_gates = self.num_gates();
+        let num_ops = gate.num_ops();
+        let gate_ref = GateRef::new(gate.clone());
+        let gate_slot = self
+            .current_slots
+            .entry(gate_ref.clone())
+            .or_insert(CurrentSlot {
```

Comment:
nit: I think it would be natural to have `CurrentSlot` `#[derive(Default)]`, and then we could replace `or_insert(...)` with `or_default()`